swarm_dial.go

```diff
@@ -281,39 +286,311 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
-// doDial is an ugly shim method to retain all the logging and backoff logic
-// of the old dialsync code
-func (s *Swarm) doDial(ctx context.Context, p peer.ID) (*Conn, error) {
-	// Short circuit.
-	// By the time we take the dial lock, we may already *have* a connection
-	// to the peer.
-	c := s.bestAcceptableConnToPeer(ctx, p)
-	if c != nil {
-		return c, nil
+///////////////////////////////////////////////////////////////////////////////////
+// lo and behold, The Dialer
+// TODO explain how all this works
+//////////////////////////////////////////////////////////////////////////////////
+type DialRequest struct {
+	Ctx   context.Context
+	Resch chan DialResponse
+}
+
+type DialResponse struct {
+	Conn *Conn
+	Err  error
+}
+
+type dialComplete struct {
+	addr ma.Multiaddr
+	conn *Conn
+	err  error
+}
+
+// dialWorker is an active dial goroutine that synchronizes and executes concurrent dials
+func (s *Swarm) dialWorker(ctx context.Context, p peer.ID, reqch <-chan DialRequest) {
+	if p == s.local {
+		for {
+			select {
+			case req, ok := <-reqch:
+				if !ok {
+					return
+				}
+
+				req.Resch <- DialResponse{Err: ErrDialToSelf}
+			}
+		}
 	}
 
-	logdial := lgbl.Dial("swarm", s.LocalPeer(), p, nil, nil)
+	s.dialWorkerLoop(ctx, p, reqch)
+}
 
-	// ok, we have been charged to dial! let's do it.
-	// if it succeeds, dial will add the conn to the swarm itself.
-	defer log.EventBegin(ctx, "swarmDialAttemptStart", logdial).Done()
+func (s *Swarm) dialWorkerLoop(ctx context.Context, p peer.ID, reqch <-chan DialRequest) {
+	defer s.limiter.clearAllPeerDials(p)
 
-	conn, err := s.dial(ctx, p)
-	if err != nil {
-		conn := s.bestAcceptableConnToPeer(ctx, p)
-		if conn != nil {
-			// Hm? What error?
-			// Could have canceled the dial because we received a
-			// connection or some other random reason.
-			// Just ignore the error and return the connection.
-			log.Debugf("ignoring dial error because we already have a connection: %s", err)
-			return conn, nil
+	type pendRequest struct {
+		req   DialRequest               // the original request
+		err   *DialError                // dial error accumulator
+		addrs map[ma.Multiaddr]struct{} // pending addr dials
+	}
+
+	type addrDial struct {
+		ctx      context.Context
+		conn     *Conn
+		err      error
+		requests []int
+	}
+
+	reqno := 0
+	requests := make(map[int]*pendRequest)
+	pending := make(map[ma.Multiaddr]*addrDial)
+
+	var triggerDial <-chan time.Time
+	var nextDial []ma.Multiaddr
+	active := 0
+	done := false
+
+	resch := make(chan dialComplete)
+
+loop:
+	for {
+		select {
+		case req, ok := <-reqch:
+			if !ok {
+				// request channel has been closed, wait for pending dials to complete
+				if active > 0 {
+					done = true
+					reqch = nil
+					triggerDial = nil
+					continue loop
+				}
+
+				// no active dials, we are done
+				return
+			}
+
+			c := s.bestAcceptableConnToPeer(req.Ctx, p)
+			if c != nil {
+				req.Resch <- DialResponse{Conn: c}
+				continue loop
+			}
```

Comment:
Let's test this one. It shouldn't be _too_ hard.

Reply:
done.